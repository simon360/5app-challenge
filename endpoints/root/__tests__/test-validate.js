const validate = require("../validate");

const getSample = () => require("./samples/request.json");

const createItem = () => ({ name: "test item", count: 4, logos: [] });
const createLogo = () => ({ url: "http://www.example.com/", size: "16x16" });

describe("validate()", () => {
  it("should succeed on provided sample", () => {
    expect(validate(getSample())).toMatchSnapshot();
  });

  it("should succeed using generated sample", () => {
    expect(validate({ payload: [createItem()] })).toMatchSnapshot();
  });

  it("should fail on missing payload", () => {
    expect(validate({})).toMatchSnapshot();
  });

  it("should fail on missing item name", () => {
    const item = createItem();

    delete item.name;

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when item name is not a string", () => {
    const item = createItem();

    item.name = 152;

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail on missing item count", () => {
    const item = createItem();

    delete item.count;

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when count is not a number", () => {
    const item = createItem();

    item.count = "4";

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail with missing logos", () => {
    const item = createItem();

    delete item.logos;

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail with non-array logos", () => {
    const item = createItem();

    item.logos = "test";

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should succeed with valid logo specified", () => {
    const item = createItem();
    const logo = createLogo();

    item.logos = [logo];

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when logo url is missing", () => {
    const item = createItem();
    const logo = createLogo();

    delete logo.url;

    item.logos = [logo];

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when logo url is not a string", () => {
    const item = createItem();
    const logo = createLogo();

    logo.url = 1412;

    item.logos = [logo];

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when logo size is missing", () => {
    const item = createItem();
    const logo = createLogo();

    delete logo.size;

    item.logos = [logo];

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when logo size is not a string", () => {
    const item = createItem();
    const logo = createLogo();

    logo.size = 16;

    item.logos = [logo];

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });

  it("should fail when logo size does not match expected format", () => {
    const item = createItem();
    const logo = createLogo();

    logo.size = "not a size";

    item.logos = [logo];

    expect(validate({ payload: [item] })).toMatchSnapshot();
  });
});
