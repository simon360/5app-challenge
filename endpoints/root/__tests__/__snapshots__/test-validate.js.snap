// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate() should fail on missing item count 1`] = `
Array [
  Object {
    "dataPath": ".payload[0]",
    "keyword": "required",
    "message": "should have required property 'count'",
    "params": Object {
      "missingProperty": "count",
    },
    "schemaPath": "#/properties/payload/items/required",
  },
]
`;

exports[`validate() should fail on missing item name 1`] = `
Array [
  Object {
    "dataPath": ".payload[0]",
    "keyword": "required",
    "message": "should have required property 'name'",
    "params": Object {
      "missingProperty": "name",
    },
    "schemaPath": "#/properties/payload/items/required",
  },
]
`;

exports[`validate() should fail on missing payload 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'payload'",
    "params": Object {
      "missingProperty": "payload",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`validate() should fail when count is not a number 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].count",
    "keyword": "type",
    "message": "should be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/properties/payload/items/properties/count/type",
  },
]
`;

exports[`validate() should fail when item name is not a string 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].name",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/payload/items/properties/name/type",
  },
]
`;

exports[`validate() should fail when logo size does not match expected format 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].logos[0].size",
    "keyword": "pattern",
    "message": "should match pattern \\"^[0-9]+x[0-9]+$\\"",
    "params": Object {
      "pattern": "^[0-9]+x[0-9]+$",
    },
    "schemaPath": "#/properties/payload/items/properties/logos/items/properties/size/pattern",
  },
]
`;

exports[`validate() should fail when logo size is missing 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].logos[0]",
    "keyword": "required",
    "message": "should have required property 'size'",
    "params": Object {
      "missingProperty": "size",
    },
    "schemaPath": "#/properties/payload/items/properties/logos/items/required",
  },
]
`;

exports[`validate() should fail when logo size is not a string 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].logos[0].size",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/payload/items/properties/logos/items/properties/size/type",
  },
]
`;

exports[`validate() should fail when logo url is missing 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].logos[0]",
    "keyword": "required",
    "message": "should have required property 'url'",
    "params": Object {
      "missingProperty": "url",
    },
    "schemaPath": "#/properties/payload/items/properties/logos/items/required",
  },
]
`;

exports[`validate() should fail when logo url is not a string 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].logos[0].url",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/payload/items/properties/logos/items/properties/url/type",
  },
]
`;

exports[`validate() should fail with missing logos 1`] = `
Array [
  Object {
    "dataPath": ".payload[0]",
    "keyword": "required",
    "message": "should have required property 'logos'",
    "params": Object {
      "missingProperty": "logos",
    },
    "schemaPath": "#/properties/payload/items/required",
  },
]
`;

exports[`validate() should fail with non-array logos 1`] = `
Array [
  Object {
    "dataPath": ".payload[0].logos",
    "keyword": "type",
    "message": "should be array",
    "params": Object {
      "type": "array",
    },
    "schemaPath": "#/properties/payload/items/properties/logos/type",
  },
]
`;

exports[`validate() should succeed on provided sample 1`] = `null`;

exports[`validate() should succeed using generated sample 1`] = `null`;

exports[`validate() should succeed with valid logo specified 1`] = `null`;
